### ISSUE TITLE
How can I compute the average recall for a dataset?

### ISSUE THREAD

**Issue by AndreaPi** [open-mmlab/mmsegmentation#1678]
I see that the metric used for model evaluation is mIoU, but I would also like to compute average recall. I see there is code for that:

https://mmsegmentation.readthedocs.io/en/latest/_modules/mmseg/core/evaluation/metrics.html?highlight=recall#

but I'm not sure how I can extract it. What should I change in the config file? Is it possible to log more than one evaluation metric to Tensorboard, e.g.

- mIoU
- mDice
- AR (average recall)
? Thanks

**Comment by MengzhangLI** [comment]
https://github.com/open-mmlab/mmsegmentation/blob/master/tools/test.py#L41-L46

Just add `--eval mIoU` or `--eval mDice` or other metric in `./tools/test.py`

**Comment by AndreaPi** [comment]
@MengzhangLI thanks for the comment, but it doesn't really answer my question:

1. I want to compute average recall. What's the name of this metric in `mmsegmentation`?
2. Also, I'd rather compute it during training, not just at test time. This way, I can train multiple models and compare their average recall on a validation set while they train, not just at the end. In other words, I'd like to put the average recall metric in the `config` file. How do I do that?

**Comment by MengzhangLI** [comment]
Sorry for late reply.

1. Currently `recall` is not supported as a metric in MMSegmentation. But you could get it followed [here](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/core/evaluation/metrics.py#L379).
2. Your requirement can not be met up because in MMSegmentation evaluation is ONLY adopted in test/validation phase.

**Comment by AndreaPi** [comment]
@MengzhangLI no worries:

1. ok, could I make a PR to add `precision` and `recall` to `allowed_metrics` here?
https://github.com/open-mmlab/mmsegmentation/blob/144dc3919e53d51921837e5a2a6d5ee656dccd7d/mmseg/core/evaluation/metrics.py#L359
3.  suppose I have the `workflow=[('train', 1), ('val', 1)]` in my config. Then, after each training epoch, a validation epoch is perfomed. Will the evaluation metric (e.g., `recall`, if I add it) be computed at the end of the validation epoch?

**Comment by MengzhangLI** [comment]
1. It's OK and always welcome! BTW, is `recall` usually used as evaluation in semantic segmentation? 
2. Yes. You could use `--eval mRecall` if you have successfully added `recall` in your codebase.

Here is a PR about [FScore](https://github.com/open-mmlab/mmsegmentation/pull/509), I think it may help you.

**Comment by AndreaPi** [comment]
1. Thanks! You're right, it's not commonly used. But we're using a semantic segmentation model to detect defects in images of machine parts...long story 🙂
2. perfect. Will have a look to the PR about FScore, and see what I can do.

**Comment by AndreaPi** [comment]
@MengzhangLI if I choose `mFscore` as an evaluation metric, it looks like `Precision` and `Recall` are added to `ret_metrics`:
```
       elif metric == 'mFscore':
            precision = total_area_intersect / total_area_pred_label
            recall = total_area_intersect / total_area_label
            f_value = torch.tensor(
                [f_score(x[0], x[1], beta) for x in zip(precision, recall)])
            ret_metrics['Fscore'] = f_value
            ret_metrics['Precision'] = precision
            ret_metrics['Recall'] = recall
```

so maybe I don't need a PR after all? I could just add the line

evaluation = dict(metric=['mIoU', 'mFscore']) and it should be possible to get the `Recall` as an output. Or no?

**Comment by MengzhangLI** [comment]
I suggest you follow PR of FScore I mentioned above if you want to add it.

Yes, you dont need a PR. It is more flexible if you modify this codebase locally.

But in next time when you want to add some interesting features to MMSegmentation, PR of you is always welcome!

**Comment by AndreaPi** [comment]
There seems to be a misunderstanding 😅 I wasn't suggesting to modify my codebase locally. I was saying: I don't need to make a PR, because the official MMSegmentation code already returns Precision and Recall in the `ret_metrics` variable, if one chooses `mFscore` as evaluation metric:

https://github.com/open-mmlab/mmsegmentation/blob/54bd4bdd82c7aa1c0ea7220b5a0672a8326e4134/mmseg/core/evaluation/metrics.py#L377

Maybe I'll create a running example and show you. Not todat because I'm feeling a bit sick, but soon.

**Comment by Sanster** [comment]
Example output when using `--eval= mFscore`

<img width="379" alt="image" src="https://user-images.githubusercontent.com/3998421/175852833-207fc072-d41e-4459-bf11-08d7c80436cb.png">

**Comment by AndreaPi** [comment]
@Sanster thanks! This confirms my hunch that using `--eval= mFscore` would compute also `Precision` and `Recall`. Thus I don't need to write a PR. Cool!
